File: tks/RepositoriesAdapters/src/main/kotlin/adapters/ResourceRepositoryAdapter.java
package adapters;

import data.AbstractResourceEntity;
import model.values.AccessionNumber;
import ports.secondary.ResourcePersistencePort;
import ports.secondary.ResourceSearchPort;
import mappers.ResourceMapper;
import model.Resource;
import org.jetbrains.annotations.NotNull;
import repositories.RepositoryBase;

import java.util.Objects;
import java.util.UUID;

public class ResourceRepositoryAdapter implements ResourcePersistencePort, ResourceSearchPort {
    private final RepositoryBase<AbstractResourceEntity> repository;
    private final ResourceMapper mapper;

    public ResourceRepositoryAdapter(RepositoryBase<AbstractResourceEntity> repository, ResourceMapper mapper) {
        this.repository = repository;
        this.mapper = mapper;
    }

    @Override
    public void add(@NotNull Resource resource) {
        var entity = mapper.mapDomainObjectToEntity(resource);
        repository.add(entity);
    }

    @Override
    public void save(@NotNull Resource resource) {
        var accessionNumberValue = resource.getId().getValue();
        var entity = repository.find(x -> Objects.equals(x.getId(), accessionNumberValue));
        assert entity != null; // todo prepare proper exception
        mapper.mapDomainObjectToEntity(resource, entity);
        repository.update(entity);
    }

    @Override
    public void remove(@NotNull Resource resource) {
        var accessionNumberValue = resource.getId().getValue();
        var entity = repository.find(x -> Objects.equals(x.getId(), accessionNumberValue));
        assert entity != null; // todo prepare proper exception
        repository.remove(entity);
    }

    @Override
    public Resource findById(AccessionNumber id) {
        var accessionNumberValue = id.getValue();
        var entity = repository.find(x -> Objects.equals(x.getId(), accessionNumberValue));
        return mapper.mapEntityToDomainObject(entity);
    }
}

File: tks/RepositoriesAdapters/_src/main/kotlin/repository/adapters/ResourceRepositoryAdapter.kt
package repository.adapters

import domain.model.context.library.Resource
import domain.model.values.AccessionNumber
import ports.secondary.IResourceRepositoryAdapter
import repository.data.AbstractResourceEntity
import repository.mappers.ResourceMapper
import repository.repositories.IRepository

class ResourceRepositoryAdapter(
    private val repository: IRepository<AbstractResourceEntity>,
    private val mapper: ResourceMapper
) : IResourceRepositoryAdapter {

    override fun save(resource: Resource) {
        val entity = repository.find { x: AbstractResourceEntity -> x.accessionNumber == resource.accessionNumber?.value }
        if(entity == null){
            repository.add(mapper.mapDomainObjectToEntity(resource)!!)
        }else{
            mapper.mapDomainObjectToEntity(resource, entity)
            repository.update(entity)
        }
    }

    override fun remove(resource: Resource) {
        val entity = repository.find { x: AbstractResourceEntity -> x.accessionNumber == resource.accessionNumber?.value } ?: throw Exception()
        repository.remove(entity)
    }

    override fun findByAccessionNumber(accessionNumber: AccessionNumber): Resource? {
        val entity = repository.find { x: AbstractResourceEntity -> x.accessionNumber == accessionNumber.value }
        return mapper.mapEntityToDomainObject(entity)
    }
}